### PR TITLE
Testing 61 ckb next doesnt remember settings after restart

### PR DESCRIPTION
**Pull request by frickler24**
At the suggestion of @tatokis, not only the writability of the config file itself is evaluated,
but also reasons why the file could not be created.

This means that the original functions of QSettings are not enough, but we want to save the system independency given by QSettings.

So first try is to create a test entry in the Config.
If this test entry can be read after closing and reopening the config file, the file is writable or can be created. 

This PR fixes issue #61 and replaces PR #71  which was based on master. This one is based on testing.

### CODE REVIEW

**Comment by ghost** [comment]
How can I test it?

**Comment by frickler24** [comment]
@light2yellow make a chmod 0 to ~/.config/.ckb-next
or make a chmod 400 or similar to ~/.config/.ckb-next/ckb-next.conf
or do something else with your Profile so that it is not writable
Then start ckb and a panel must appear with the warning.

If conf file is writable, nothing new happens.

**Comment by ghost** [comment]
I see the message but there are few issues I would like to point out:
* The spacing between the upper and the lower borders is... too big?
* I had to close a window asking me to choose my layout first (that window that everyone see when launching for the first time). I think it should be displayed after presing OK on the window generated by this patch

So, basically it's tested and does its job.

File: src/ckb/ckbsettings.cpp [hunk 1]
```diff
@@ -31,6 +59,42 @@ static QSettings* globalSettings(){
...
+/// \brief CkbSettings::checkIfWritable
+/// If the local implementation of the config database is not yet writable,
+/// bring up a popup to the user to informhim about it.
+/// Bring up the information where he can find the info.
+bool CkbSettings::informIfNotWritable() {
+    if (isWritable()) return false;
+    QMessageBox msgBox;
+    msgBox.setIcon(QMessageBox::Warning);
+    msgBox.setText("\n\n\nYour profile information for ckb-next is not writable.\n");
+    QString info = "This might happen, if you started the ckb-next program with root privileges earlier.\nOr did you copy it from somewhere?\n\nPlease have a look at "
```
Comment (ghost) [inline src/ckb/ckbsettings.cpp]: > This might happen, if

Comma not needed.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/ckb/ckbsettings.cpp (hunk 1 above)
Comment (ghost) [reply src/ckb/ckbsettings.cpp]: (I'm also trying out "review" features :) )

Comment (frickler24) [reply src/ckb/ckbsettings.cpp]: good function!
Now I can see my correction commit, but it is in the "normal" flow.
How can I mark  9556c84 as corresponding to your specific review?

**Comment by frickler24** [comment]
@light2yellow I reduced the spacing a bit and moved the filename into a new line.
More complex would be to synchronize the two panels, because they come from different threads and the creation of the thread is done before checking the config file.
The only chance would be with a separate semaphore.
Or is there a way to check in Qt for existence of a somehow statically named window?

File: src/ckb/ckbsettings.cpp [hunk 2]
```diff
@@ -17,11 +17,39 @@ QMutex settingsMutex(QMutex::Recursive), settingsCacheMutex(QMutex::Recursive);
...
+            ///
+            CkbSettings::setWritable(false);
+            QSettings* testSettings = new QSettings;
+            testSettings->setValue("testIfWritable", 42);
+            testSettings->sync();
+            delete testSettings;
+
+            /// Try to open the conf file again and to read the standard value. Delete the standard value afterwards.
+            /// Because on linux QSettings does somehow caching, it reads the correct value even the file (not the directory) is read only.
+            /// So we need to add the original isWritabel() function.
```
Comment (ghost) [inline src/ckb/ckbsettings.cpp]: isWritabel() <- typo

**Comment by ghost** [comment]
> ...if you did start the ckb-next program with root privileges earlier...

According to [ckb/main.cpp#L165](https://github.com/mattanger/ckb-next/blob/testing/src/ckb/main.cpp#L165) the app won't allow a user to be run as root under any circumstances. So I think this is redundant. I guess a message like 

> Your profile information for ckb-next is nonwritable `note: "not writable" -> "nonwritable"`. Please check the file system permissions of %ckb_dirname%.

should suffice.

We should also log it with `ckb_err()`.

> The only chance would be with a separate semaphore.

Isn't it possible to merge them together? I mean adding this functionality into the method that checks whether the config exists (just that), then if it does, check for being writable.

**Review by ghost — APPROVED** [review]

File: src/ckb/ckbsettings.cpp (hunk 1 above)
Comment (ghost) [reply src/ckb/ckbsettings.cpp]: I have no idea ¯\_(ツ)_/¯

**Comment by frickler24** [comment]
@light2yellow Sorry, I had a little trouble pushing my last commit. If there is a problem with the last commit  	dc02a24, I'll try to create a clean one again.

**Comment by ghost** [comment]
Can someone else test this so it can be merged?

**Comment by ghost** [comment]
Oh, I actually missed the new commits, sorry. I will re-test it tomorrow.

**Comment by tatokis** [comment]
I just had some time to try this.
It still doesn't detect if the directory is owned by root.

Additionally, I think we should quit after the user has seen the warning, instead of normally resuming operation, in order to avoid any possible profile overwriting.

**Comment by fleischie** [comment]
Hey, I have a mac now (thank you employer 😀 ). Is there anything mac-related I should test? Respectively: What's the status on this?

**Comment by frickler24** [comment]
@tatokis Now I added an explicit separate file for checking your condition mentioned.
Did you delete your config file when you chown-ed the directory to root? Otherwise there is a file which you own in a directory which is not writable - but the file is writable. 
Both cases were detected correctly on my system. Nevertheless, now a writable file in a non writable directory is detected and mentioned.
As you wanted, the client stops now after the message disappeared.
@fleischie Please test it with your new mac (congratulation for your employer :-) 
Find any situation where you cannot access your config file:
File is r/o, you do not own the file, you do not own the directory with the file and have no permission to write to the dir etc.
After all tests there should be no new additional file in the directory where your configs are stored.

**Comment by fleischie** [comment]
Hey, I have done some testing on Mac OS (El Capitan 10.11.6).

<details>
First of all, it took me like 2 hours to find the settings file. 😅
Edit: And `ckb`'s first console line states the path of the settings file. :man_facepalming: 
</details>

# Tests

## Settings file does not exist (prior to test)

### Directory permission permutations

| dir readable | dir writeable | dir executable | result |
| --- | --- | --- | --- |
| true | true | true | default behavior |
| true | true | false | open dialog |
| true | false | true | open dialog |
| true | false | false | open dialog |
| false | true | true | default behavior |
| false | true | false | open dialog |
| false | false | true | open dialog |
| false | false | false | open dialog |

Conclusion: If directory is writeable and executable the default behavior is triggered, in any other case the dialog box is opened. Strangely `ckb` is not asking for the keyboard layout.

## Settings file exists

### File permission permutations

Prerequisite: Directory is 700 with karl:staff ownership.

| file readable | file writeable | result |
| --- | --- | --- |
| true | true | default behavior |
| true | false | default behavior |
| false | true | initial start |
| false | false | initial start |

Conclusion: File readability determines, whether the default or the initial behavior is triggered.

### Dir permission permutations

Prerequisite: File mode is 400 and has karl:staff ownership.

| dir readable | dir writeable | dir executable | result |
| --- | --- | --- | --- |
| true | true | true | default behavior |
| true | true | false | open dialog |
| true | false | true | open dialog |
| true | false | false | open dialog |
| false | true | true | default behavior |
| false | true | false | open dialog |
| false | false | true | open dialog |
| false | false | false | open dialog |

Conclusion: If directory is writeable and executable the default behavior is triggered, in any other case the dialog box is opened.

----

I did some more permutations with file/dir permissions + ownerships but it boils down to the above-mentioned tables.

Additionally I don't think the operating system flavor (mac vs. linux) has an effect on this table, as the rwx-permissions are pretty much the same and it's all in user-spaced territory (the settings file is not stored on global directories).

P.S.: I might have over-engineered the tests for such a simple PR. 😄 

Edit: Do you need me to review the code as well? I just unit tested the input-output of the PR.

**Comment by frickler24** [comment]
@fleischie Wow, thanks a lot!
To 1.: 
> Strangely ckb is not asking for the keyboard layout.

This is correct, Tatokis tip was to terminate the session if config file is nonwritable.
To 2.:
> true 	false 	default behavior: 

This should not happen. On my linux I get the open dialog.
Please have a look at ckbsettings.cpp: static QSettings* globalSettings(){ (beginning from line 27).
At least the function testSettings->isWritable() in line 64 should give false as retval.
Your third matrix shows the same effect. The dependency should both be from write-access and read-acces.
perm==0--> Warning-panel
perm==400--> Warning-panel
perm==200--> currently I did not check this one. Behavior at linux is Warning-panel because of testSettings->isWritable()==false.
perm==600--> default

My thought was to use the machine independent function as far as possible.
If we get problems with it, I change it and have a look at fstat at the dir and file.

**Pull request closed without merging**